core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java

```diff
@@ -314,14 +315,36 @@ public void onComplete() {
                 responseEncoder.writeTrailers(req.id(), req.streamId(), additionalTrailers)
                                .addListener(writeHeadersFutureListener(true));
                 ctx.flush();
-            } else if (responseEncoder.isWritable(req.id(), req.streamId())) {
-                responseEncoder.writeData(req.id(), req.streamId(), HttpData.empty(), true)
-                               .addListener(writeDataFutureListener(true, true));
-                ctx.flush();
+            } else {
```

Comment:
> Question) I thought the following lines existed specifically to address this problem (although the exception check is failing at the moment due to ClosedSessionException being passed instead of ClosedChannelException)

I didn't know that `WriteDataFutureListener` has such a workaround. It might be a good idea to delegate the error handling to `WriteDataFutureListener` for a clean implementation. Although, it would be a little bit inefficient compared to the current one.

> Also, I think the previous implementation also handles the following case where an empty data is passed explicitly. Is there need to handle this case as well?

So now I am thinking of handling `ClosedSessionException` in `WriteDataFutureListener`. You convinced me.

Reply:
I didn't know this logic existed 😅    Makes sense  👍 
Then I guess it makes sense to have two places for this logic since 1) we don't know if a write will fail or not 2) it's more efficient to just bail out earlier if possible